text/0000-string-reflection.md

```diff
@@ -0,0 +1,148 @@
+  if one is available;
+- get the owned string's content if the type matches.
+
+# Motivation
+[motivation]: #motivation
+
+The trait `Any` is Rust's way to represent arbitrary values (of `'static`
+types) while enabling runtime type checking and downcasting. One of the most
+prominent uses of this trait, with a `Box<Any>` object pointer, is to convey
+the value of the `panic!` invocation retrieved from a panicked thread.
+The `Box<Any>` is returned by `std::thread::JoinHandle::join()` in the
+`Err` variant. Calling code that does not expect the child thread to panic
+often uses the `unwrap()` method of the `Result` value, so that, if the child
+thread has panicked, the calling thread would panic in turn. The generic
+implementation of `Result::unwrap()` uses the `Debug` implementation of the
+`Err`'s value for content of its own panic message. However, the current
+`Debug` implementation for `Any` trait objects [does not perform][rfc-issue]
+any reflection on the value, therefore disrupting propagation of the
+originating panic's message even if it is a string.
+
+This leaves the panic hook as the only way to inform the user about the cause
+of the originating panic. The process-global panic hook can be overridden
+for unrelated reasons, leaving potential for unintended loss of information
+on errors within the program.
+
+In other uses, it would also be nice and informative to have the `Debug` impl
+output the incidental string value under an `Any` trait object, while not
+behaving differently for owned and static strings as long as they are
+represented with their standard Rust types.
+
+The problem with implementing it in this way is that a consistent
+implementation would have to include a check for the `String` type,
+which is not known to libcore where the implementation is defined.
+
+[rfc-issue]: https://github.com/rust-lang/rfcs/issues/1389
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Debug` trait implementation for trait object type `Any` (and for
+`Any + Send`) accommodates a special case: if the actual type of the value
+is `&'static str` or `String` (temporary references or unsized slices can't
+be under `Any`), the string content is formatted.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The main issue to solve here is how to make the impl code in libcore
+recognize the type of an `Any` object value as `String` and get a peek
+into its content, even as this allocating type is defined above libcore in the
+dependency graph.
+
+## String reflection protocol
+[api]: #string-reflection-protocol
+
+Libcore implements the dynamic type check for, and downcasting to the content
+of, the owning string type if such exists in the linkage, a couple of external
+symbols that are linked with the implementation crate by the compiler,
+defined approximately as:
+
+```rust
+extern {
+    fn __rust_type_is_string(id: TypeId) -> bool;
```

Comment:
Why isn't `id == TypeId::of::<String>()` sufficient?

Reply:
Yes, it would be trivial if not for the need for a lazy-linked dependency.